blockchain/chain_sync/src/tipset_syncer.rs

```diff
@@ -187,21 +187,31 @@ impl TipsetGroup {
     }
 
     fn take_heaviest_tipset(&mut self) -> Option<Arc<Tipset>> {
-        self.tipsets
-            .iter()
-            .enumerate()
-            .max_by_key(|(_idx, ts)| ts.weight())
-            .map(|(idx, _)| idx)
-            .map(|idx| self.tipsets.swap_remove(idx))
+        let (index, _) = self.heaviest_weight();
+        Some(self.tipsets.swap_remove(index))
     }
 
-    fn heaviest_weight(&self) -> BigInt {
+    fn heaviest_weight(&self) -> (usize, &BigInt) {
         // Unwrapping is safe because we initialize the struct with at least one tipset
-        self.tipsets
+        let max = self.tipsets.iter().map(|ts| ts.weight()).max().unwrap();
+
+        let ties = self
+            .tipsets
             .iter()
-            .map(|ts| ts.weight().clone())
-            .max()
-            .unwrap()
+            .enumerate()
+            .filter(|(_, ts)| ts.weight() == max);
+
+        let (index, ts) = ties
+            .reduce(|(i, ts), (j, other)| {
+                // break the tie
+                if ts.break_weight_tie(other) {
```

Comment:
Do you think we could make it more idiomatic by implementing `PartialOrd` for `Tipset` with this logic? Seems to me this logic should be contained there, I may be in the wrong though. Thoughts?